[Chua Bing Quan] iP

chuabingquan opened the pull request:
# Duke 🤡
> "Manage your tasks like a nobility." - Einstein (probably)

## Why Duke?
- Who doesn't clamour for an app that enhances our micromanagement capabilities? 🤡
- Profit??? 💸

## Features
Manage the ~~burdens~~ joys of life by
1. Creating new tasks (e.g. `Todo`, `Deadline`, `Event`)
2. Viewing all tasks 😲
3. Finding tasks with a keyword 😧
4. Marking tasks as completed 🤭
5. Marking tasks as uncompleted 😵
6. Deleting tasks 😱
7. Exiting ~~and giving up on your tasks~~ 🤯

## Other features you definitely care about
It's written in Java™! Yes, you've heard it right, the one and [only](https://www.johndcook.com/blog/2011/07/19/you-wanted-banana/) [language](https://preview.redd.it/rqsz6jzvfif11.png?auto=webp&s=96e864fac0e5066f5e0f72cb819e24ad8dda583c) [for](https://www.reddit.com/r/ProgrammerHumor/comments/efzvct/i_oop/) [enterprise™](https://github.com/EnterpriseQualityCoding/FizzBuzzEnterpriseEdition/) [systems](https://files.speakerdeck.com/presentations/9c6227be4d424916bdc63795e1b1f4af/slide_4.jpg)! Just check out some of its greatest hits:
- [x] Null pointer exceptions
- [x] Generics implemented with type erasure 
- [x] Inheritance (for long inheritance chains of course!)
- [x] And many more, google them!
  
Also check out the number of lines of code it takes to define a linked list in Java™:
```java
// Define Node class...

// Define List class...
class List<T> {
    private Node<T> head;
    private List<T> tail;

    public List() {
        this.head = null;
        this.tail = null;
    }

    public List(T item) {
        this.head = item;
        this.tail = null;
    }

    // a lot more boilerplate awaits you!
}
```
Now look at how *weak* Haskell is in comparison:
```hs
data List a = Empty | Pair a (List a)
```
That future boss of yours who measures productivity by `kloc` would be absolutely __*floored*__! 🤡 🤡 🤡

## Disclaimer
This is a joke, in case you can't tell.

Commit:
Add Gradle support

chuabingquan committed:
Implement greet, echo, and exit

chuabingquan committed:
Implement add and list commands

chuabingquan committed:
Refactor stored items as Task objects

chuabingquan committed:
Rename items array to reflect object type

chuabingquan committed:
Refactor duke commands to correspond to relevant methods

chuabingquan committed:
Show status icons for list of tasks

chuabingquan committed:
Implement marking and unmarking of tasks

chuabingquan committed:
Organise Task into its own package

chuabingquan committed:
Add task variants

chuabingquan committed:
Fix incorrect word

chuabingquan committed:
Enable creation of task variants

chuabingquan committed:
Fix bug where task status indicators are shown incorrectly

chuabingquan committed:
Fix bug where adding task reflects the wrong number of existing tasks

chuabingquan committed:
Allow task subclasses to access getStatusIcon

chuabingquan committed:
Rename instances of ToDo to Todo

chuabingquan committed:
Implement basic UI testing

chuabingquan committed:
Shift bot logic into its own class

chuabingquan committed:
Update tests to account for some errors

chuabingquan committed:
Implement exceptions for empty task descriptions and unknown commands

chuabingquan committed:
Refactor bot to use array list for tasks

chuabingquan committed:
Implement task deletion

chuabingquan committed:
Add tests for task deletion

chuabingquan committed:
Create types for TaskService

chuabingquan committed:
Allow cloning of tasks

chuabingquan committed:
Update TaskService interface to support exceptions

chuabingquan committed:
Implement task retrieval from file system

chuabingquan committed:
Implement create, update, and deletion of tasks to/from file system

chuabingquan committed:
Remove unused import

chuabingquan committed:
Define and implement get number of tasks

chuabingquan committed:
Fix bug where getAt goes into an infinite loop

chuabingquan committed:
Fix bug where task service tries to read/write to restricted storage

chuabingquan committed:
Refactor bot to use task service for task-related CRUD operations

chuabingquan committed:
Remove auto-generated test files

chuabingquan committed:
Throw relevant errors from file system task service

chuabingquan committed:
Rename prefix to adhere to the global variable naming convention

chuabingquan committed:
Use date instead of string to represent deadlines

chuabingquan committed:
Merge branch 'branch-Level-7'
Implement persistence for tasks using the file system

chuabingquan committed:
Merge branch 'branch-Level-8'

Modify deadline to work with dates instead of strings

BrendonLau commented on src/main/java/bot/Bot.java:
```diff
@@ -0,0 +1,169 @@
+    private String process(String query) throws BotException {
+        try {
+            final String[] tokens = query.split(" ", 2);
+            switch (tokens[0]) {
```
I think I have the same problem, maybe for readability, can assign meaning to 0,1

BrendonLau commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,43 @@
+        return this.isDone ? "X" : " ";
+    }
+
+    public boolean getIsDone() {
```
For the naming convention, I think isDone() is sufficient and get no needed.

BrendonLau commented on src/main/java/taskservice/FileSystemTaskService.java:
```diff
@@ -0,0 +1,179 @@
+            throw new TaskServiceException(FileSystemTaskService.EXCEPTION_INVALID_FILE_FORMAT);
+        }
+
+        Task task;
+        switch (tokens[0].trim()) {
+        case "T":
+            task = new Todo(tokens[2].trim());
+            break;
+        case "D":
+            task = new Deadline(tokens[2].trim(), LocalDate.parse(tokens[3].trim(), Deadline.DATE_INPUT_FORMAT));
+            break;
+        case "E":
+            task = new Event(tokens[2].trim(), tokens[3].trim());
+            break;
+        default:
+            throw new TaskServiceException(FileSystemTaskService.EXCEPTION_INVALID_FILE_FORMAT);
```
Similar as above, for readability may be good to assign a variable to token[0/1/2].

BrendonLau reviewed (commented):
Overall, LGTM. Good job!

snss231 commented on src/main/java/bot/Bot.java:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    private String unmarkTask(int taskId) throws BotException, TaskServiceException {
+        final Optional<Task> result = this.taskService.getById(taskId);
```
Maybe it would be better if the name "result" was changed to something more explicit

snss231 commented on src/main/java/taskservice/FileSystemTaskService.java:
```diff
@@ -0,0 +1,179 @@
+    private static final String TASKS_FILENAME = "tasks";
+    private static final String EXCEPTION_INVALID_FILE_FORMAT = "Unable to parse invalid tasks file from file system";
+
+    private final Path rootDir;
```
Consts should be named using screaming snake case

snss231 reviewed (commented):
Couldn't really find any coding standard violations other than the final thingy, really polished submission congratz

ianfromdover commented on src/main/java/tasks/Todo.java:
```diff
@@ -0,0 +1,21 @@
+        super(description, isDone);
+    }
+
+    @Override
+    public Todo clone() {
+        return new Todo(super.description, super.isDone);
+    }
```
Interesting method, perhaps a header comment would be helpful for me to understand why you included it :)

ianfromdover commented on src/main/java/taskservice/FileSystemTaskService.java:
```diff
@@ -0,0 +1,179 @@
+            this.hasReadFromFileSystem = true;
+        }
+
+        return this.tasks.stream().map((task) -> task.clone()).toArray(Task[]::new);
```
Innovative use of streams here. Good job 👍

ianfromdover commented on src/main/java/bot/Bot.java:
```diff
@@ -0,0 +1,169 @@
+        }
+    }
+
+    private String process(String query) throws BotException {
```
Perhaps this can be broken into a `Parser` class :)

ianfromdover reviewed (commented):
Well organised and follows coding standards closely. Innovative use of streams 👍

chuabingquan committed:
Extract UI functionality into its own class

chuabingquan committed:
Expose showDeleteTask as a public method

chuabingquan committed:
Implement UI for showing errors

chuabingquan committed:
Rename parameters for clarity

chuabingquan committed:
Allow UI to get user input

chuabingquan committed:
Implement serialization and deserialization of tasks file

chuabingquan committed:
Implement loading and saving of tasks from/to file-system

chuabingquan committed:
Implement CRUD operations for tasks from file-system

chuabingquan committed:
Introduce command abstraction to handle command execution

chuabingquan committed:
Implement parser to validate and route user commands

chuabingquan committed:
Remove message prefix

chuabingquan committed:
Complete program REPL

chuabingquan committed:
Remove task service implementation

chuabingquan committed:
Merge remote-tracking branch 'origin/add-gradle-support'
Add gradle support

chuabingquan committed:
Add unit tests for Deadline

chuabingquan committed:
Update main class name in gradle config

chuabingquan committed:
Write javadocs for the bot package

chuabingquan committed:
Write javadocs for the parser package

chuabingquan committed:
Include javadocs for the storage package

chuabingquan committed:
Include javadocs for the tasklist package

chuabingquan committed:
Include javadocs for the tasks package

chuabingquan committed:
Remove redundant spacing in javadocs

chuabingquan committed:
Include javadocs for the commands package

chuabingquan committed:
Include javadocs for the main entry point to the chat bot

chuabingquan committed:
Fix spelling issue in javadocs

chuabingquan committed:
Include javadocs for the ui package

chuabingquan committed:
Specify javadocs for class members on a single line

chuabingquan committed:
Remove extra semicolon

chuabingquan committed:
Implement the ability to find tasks by keyword

chuabingquan committed:
Implement the find command

chuabingquan committed:
Fix incorrect grouping of imports

chuabingquan committed:
Implement the UI for finding tasks by keyword

chuabingquan committed:
Merge branch 'branch-A-JavaDoc'
Include javadocs for all publicly exported types

chuabingquan committed:
Merge branch 'branch-A-CodingStandard'
Refactor program to conform with style guide.

chuabingquan committed:
Merge branch 'branch-Level-9'
Complete implementation of find tasks by keyword

chuabingquan committed:
Configure checkstyle

chuabingquan committed:
Fix code to comply checkstyle rules

chuabingquan committed:
Merge branch 'branch-A-CheckStyle'
Add checkstyle config and fix code to comply with checkstyle rules

chuabingquan committed:
Allow chatbot to execute a single query

chuabingquan committed:
Add GUI assets

chuabingquan committed:
Add JavaFX as a dependency

chuabingquan committed:
Implement basic GUI for duke

chuabingquan committed:
Add start command for bot to greet user

chuabingquan committed:
Implement in-memory Duke chatbot

chuabingquan committed:
Start GUI with a greeting

chuabingquan committed:
Allow GUI to exit on relevant commands

chuabingquan committed:
Include javadocs for the Duke GUI

chuabingquan committed:
Fix bug where dialog gets truncated

chuabingquan committed:
Merge branch 'branch-Level-10'
Implement a GUI for the Duke chatbot

chuabingquan committed:
Improve ease of debugging

Assertions are included to test for invariants and pre-conditions. This
way, any unexpected assertion violations would swiftly be made known,
easing the debugging process when modifying the program in the future.

chuabingquan committed:
Refactor program to enhance code quality

Improvements to code quality has been made to better describe their
semantics and delineate them by their logical blocks. With these
enhancements, any future maintenance to this code base can be made with
ease and enable productive collaboration.

chuabingquan committed:
Merge pull request #2 from chuabingquan/branch-A-Assertions

Improve ease of debugging

chuabingquan committed:
Merge branch 'master' into branch-A-CodeQuality
Bring the current branch up to date with master

chuabingquan committed:
Merge pull request #3 from chuabingquan/branch-A-CodeQuality

Refactor program to enhance code quality

chuabingquan committed:
Add priority to tasks

chuabingquan committed:
Implement getting the priority of a task

chuabingquan committed:
Implement string to task priority conversion

chuabingquan committed:
Support formatting of task priority to/from data file

chuabingquan committed:
Update UI to support changing of task priority

chuabingquan committed:
Implement command to enable changing of task priority

chuabingquan committed:
Merge branch 'branch-C-Priority'
Implement task priority feature

chuabingquan committed:
Make dialog display picture circular

chuabingquan committed:
Add title for GUI

chuabingquan committed:
Add delay to bot response for realism

chuabingquan committed:
Fix minor code style inconsistency

chuabingquan committed:
Merge branch 'branch-A-BetterGui'
Improve GUI user experience
* Display pictures are now circular to appear more welcoming
* Duke pauses for a moment before responding to user
* GUI closes less abruptly upon exit

chuabingquan committed:
Switch default data file directory

chuabingquan committed:
Include product image for product website

chuabingquan committed:
Update user guide

chuabingquan committed:
Update quick start with priority change example

chuabingquan committed:
Set theme jekyll-theme-cayman

chuabingquan committed:
Add the `find` command to the feature list
